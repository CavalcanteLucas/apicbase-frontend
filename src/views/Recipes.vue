<template>
  <div id="recipes">
    <div class="container">
      <div class="col-12 mx-auto">
        <div class="row">
          <div class="col-4 mx-auto">
            <div class="input-group mb-3">
              <span class="input-group-text" id="search-recipe"
                ><img src="../assets/img/search.svg" alt="Search"
              /></span>
              <input type="text" id="searchField" class="form-control" v-model="searchQuery" />
              <button
                class="btn btn-outline-secondary"
                type="button"
                id="clearSearchBtn"
                @click="clearSearchField"
              >
                <img src="../assets/img/backspace.svg" alt="Clear Search" />
              </button>
            </div>
          </div>
        </div>

        <div class="row" style="padding-top: 1rem;">
          <div class="col-6 mx-auto">
            <form id="recipesForm" @submit.prevent="submitForm">
              <div class="input-group">
                <input
                  type="text"
                  class="form-control"
                  id="inputName"
                  placeholder="Name"
                  aria-label="Name"
                  v-model="recipe.name"
                  required="required"
                />
                <button
                  type="button"
                  id="clearBtn"
                  v-on:click="clearRecipeForm()"
                  class="btn btn-outline-secondary"
                >
                  <img src="../assets/img/backspace.svg" alt="Clear" />
                </button>
                <button type="submit" id="submitBtn" class="btn btn-primary">
                  Save
                </button>
              </div>
            </form>
          </div>
        </div>

        <p class="error" style="padding-top: 1rem;" v-if="!isResponseOK">
          Something went wrong. Please try again..
        </p>

        <div class="row" style="padding-top: 2rem;">
          <div class="col-8 mx-auto">
            <table
              class="table table-striped table-hovering"
              aria-label="Recipes table"
            >
              <thead>
                <th scope="col"></th>
              </thead>
              <tbody>
                <tr
                  v-for="recipe in resultQuery"
                  :key="recipe.id"
                  @click="$data.recipe = recipe"
                >
                  <td>
                    <button
                      type="button"
                      id="detailBtn"
                      class="btn btn-sm btn-dark"
                      @click="viewRecipeDetail(recipe.id)"
                    >
                      <img src="../assets/img/eye.svg" alt="Details" />
                    </button>
                  </td>
                  <td class="align-middle">{{ recipe.name }}</td>
                  <td>
                    <button
                      type="button"
                      id="deleteBtn"
                      class="btn btn-sm btn-danger"
                      @click="deleteRecipe(recipe)"
                    >
                      <img src="../assets/img/trash.svg" alt="Delete" />
                    </button>
                  </td>
                  <confirm-dialogue ref="confirmDialogue"></confirm-dialogue>
                </tr>
              </tbody>
            </table>
          </div>
        </div>
      </div>
    </div>
  </div>
</template>

<script>
import ConfirmDialogue from "../components/ConfirmDialogue.vue";

export default {
  name: "Recipes",
  components: { ConfirmDialogue },
  data() {
    return {
      searchQuery: null,
      recipe: {
        name: null,
      },
      recipes: [],
      isResponseOK: true,
    };
  },
  async created() {
    await this.getRecipes();
  },
  methods: {
    submitForm() {
      if (this.recipe.id === undefined) {
        this.createRecipe();
      } else {
        this.editRecipe();
      }
    },
    async getRecipes() {
      let response = await fetch("http://localhost:8000/api/recipes/");
      this.recipes = await response.json();
    },
    async createRecipe() {
      await this.getRecipes();
      let response = await fetch("http://localhost:8000/api/recipes/", {
        method: "post",
        headers: {
          "Content-Type": "application/json",
        },
        body: JSON.stringify(this.recipe),
      });
      this.isResponseOK = response.ok;
      await this.getRecipes();
    },
    async editRecipe() {
      await this.getRecipes();
      await fetch(`http://localhost:8000/api/recipes/${this.recipe.id}/`, {
        method: "put",
        headers: {
          "Content-Type": "application/json",
        },
        body: JSON.stringify(this.recipe),
      });
      await this.getRecipes();
    },
    async deleteRecipe() {
      await this.getRecipes();
      const ok = await this.$refs.confirmDialogue[0].show();
      if (ok) {
        await fetch(`http://localhost:8000/api/recipes/${this.recipe.id}/`, {
          method: "delete",
          headers: {
            "Content-Type": "application/json",
          },
          body: JSON.stringify(this.recipe),
        });
      }
      await this.getRecipes();
    },
    viewRecipeDetail(recipeId) {
      this.$router.push({ name: "RecipeDetails", params: { id: recipeId } });
    },
    clearRecipeForm() {
      this.recipe = {
        name: null
      };
      this.isResponseOK = true;
    },
    clearSearchField() {
      this.searchQuery = "";
    },
  },
  computed: {
    resultQuery() {
      if (this.searchQuery) {
        return this.recipes.filter((recipe) => {
          return this.searchQuery
            .toLowerCase()
            .split(" ")
            .every((item) => {
              return recipe.name.toLowerCase().includes(item);
            });
        });
      } else {
        return this.recipes;
      }
    },
  },
};
</script>

<style></style>
