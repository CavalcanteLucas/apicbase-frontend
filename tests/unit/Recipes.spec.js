import { mount, config } from "@vue/test-utils";
import Recipes from "../../src/views/Recipes.vue";

import { columnData } from "../utils.js";

config.showDeprecationWarnings = false;

describe("recipes view", () => {
  let wrapper;
  beforeEach(() => {
    wrapper = mount(Recipes, {
      methods: {
        getRecipes: () => {
          // do nothing
        },
      },
    });
  });

  it("renders", () => {
    expect(wrapper.exists()).toBe(true);
    expect(wrapper).toMatchSnapshot();
  });

  it("has data", () => {
    expect(typeof Recipes.data).toBe("function");
  });

  it("is a Vue instance", () => {
    expect(wrapper.isVueInstance()).toBeTruthy();
  });

  it("has one input field, and one search field", () => {
    expect(wrapper.contains('input[id="inputName"]')).toBe(true);
    expect(wrapper.contains('input[id="searchField"]')).toBe(true);
  });

  it("has a clear input button, a clear search button, and a submit button", () => {
    expect(wrapper.contains('button[id="clearBtn"]')).toBe(true);
    expect(wrapper.contains('button[id="submitBtn"]')).toBe(true);
    expect(wrapper.contains('button[id="clearSearchBtn"]')).toBe(true);
  });

  it("input fields load data", () => {
    expect(wrapper.vm.$data.recipe).toEqual({
      name: null,
    });
    wrapper.find('input[id="inputName"]').setValue("cake-test");
    expect(wrapper.find('input[id="inputName"]').element.value).toBe(
      "cake-test"
    );
    expect(wrapper.vm.$data.recipe).toEqual({
      name: "cake-test",
    });
  });

  it("clear button wipes loaded data out", async () => {
    await wrapper.setData({
      recipe: {
        name: "cake-test",
      },
    });

    expect(wrapper.vm.$data.recipe).toEqual({
      name: "cake-test",
    });

    let clearBtn = wrapper.find('button[id="clearBtn"]');
    expect(clearBtn.exists()).toBe(true);
    expect(clearBtn.is("button")).toBe(true);
    await clearBtn.trigger("click");

    expect(wrapper.vm.$data.recipe).toEqual({
      name: null,
    });
    expect(wrapper.find('input[id="inputName"]').element.value).toBe("");
  });

  it("opens confirmation modal when delete button is pressed", async () => {
    expect(wrapper.findAll("tbody > tr").wrappers).toHaveLength(0);
    await wrapper.setData({
      recipes: [
        {
          name: "cake-sample",
        },
      ],
    });

    expect(wrapper.findAll("tbody > tr").wrappers).toHaveLength(1);
    let deleteBtn = wrapper.find('button[id="deleteBtn"]');
    expect(deleteBtn.exists()).toBe(true);
    expect(deleteBtn.is("button")).toBe(true);
    let modal = wrapper.find('div[class="popup-modal"]');
    expect(modal.exists()).toBe(false);

    await deleteBtn.trigger("click");

    modal = wrapper.find('div[class="popup-modal"]');
    expect(modal.exists()).toBe(true);
    expect(modal.is("div")).toBe(true);
    expect(modal.text()).toContain("Are you sure about that?");
    expect(modal.text()).toContain("Nevermind..");
    expect(modal.text()).toContain("Just do it!");
  });

  it("closes confirmation modal when cancel button is pressed", async () => {
    await wrapper.setData({
      recipes: [
        {
          name: "carrot-sample",
        },
      ],
    });
    let deleteBtn = wrapper.find('button[id="deleteBtn"]');
    expect(deleteBtn.exists()).toBe(true);
    expect(deleteBtn.is("button")).toBe(true);
    await deleteBtn.trigger("click");

    let cancelBtn = wrapper.find('button[id="cancelBtn"]');
    expect(cancelBtn.exists()).toBe(true);
    expect(cancelBtn.is("button")).toBe(true);
    await cancelBtn.trigger("click");

    let modal = wrapper.find('div[class="popup-modal"]');
    expect(modal.exists()).toBe(false);
    cancelBtn = wrapper.find('button[id="cancelBtn"]');
    expect(cancelBtn.exists()).toBe(false);
  });

  it("clear search button wipes searched data out", async () => {
    let searchField = wrapper.find('input[id="searchField"]');
    expect(searchField.exists()).toBe(true);
    expect(searchField.is("input")).toBe(true);
    await searchField.setValue("sample recipe");
    expect(searchField.element.value).toBe("sample recipe");

    let clearSearchBtn = wrapper.find('button[id="clearSearchBtn"]');
    expect(clearSearchBtn.exists()).toBe(true);
    expect(clearSearchBtn.is("button")).toBe(true);
    await clearSearchBtn.trigger("click");
    expect(searchField.element.value).toBe("");
  });

  it("filters searched data", async () => {
    expect(wrapper.findAll("tbody > tr").wrappers).toHaveLength(0);
    await wrapper.setData({
      recipes: [
        {
          name: "sample recipe X",
        },
        {
          name: "sample recipe Y",
        },
      ],
    });
    let rows = wrapper.findAll("tbody > tr").wrappers;
    expect(rows).toHaveLength(2);

    // filters sample-X by name
    let searchField = wrapper.find('input[id="searchField"]');
    expect(searchField.exists()).toBe(true);
    expect(searchField.is("input")).toBe(true);
    await searchField.setValue("x");
    rows = wrapper.findAll("tbody > tr").wrappers;
    expect(rows).toHaveLength(1);
    expect(columnData(1, rows)[0]).toBe("sample recipe X");

    // filters sample-Y by name
    await searchField.setValue("y");
    rows = wrapper.findAll("tbody > tr").wrappers;
    expect(rows).toHaveLength(1);
    expect(columnData(1, rows)[0]).toBe("sample recipe Y");
  });
});
