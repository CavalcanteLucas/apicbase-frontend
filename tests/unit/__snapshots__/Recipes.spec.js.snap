// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`recipes view renders 1`] = `
<div id="recipes">
  <div class="container">
    <div class="col-12 mx-auto">
      <div class="row">
        <div class="col-4 mx-auto">
          <div class="input-group mb-3"><span id="search-recipe" class="input-group-text"><img src="../assets/img/search.svg" alt="Search"></span> <input type="text" id="searchField" class="form-control"> <button type="button" id="clearSearchBtn" class="btn btn-outline-secondary"><img src="../assets/img/backspace.svg" alt="Clear Search"></button></div>
        </div>
      </div>
      <div class="row" style="padding-top: 1rem;">
        <div class="col-6 mx-auto">
          <form id="recipesForm">
            <div class="input-group"><input type="text" id="inputName" placeholder="Name" aria-label="Name" required="required" class="form-control"> <button type="button" id="clearBtn" class="btn btn-outline-secondary"><img src="../assets/img/backspace.svg" alt="Clear"></button> <button type="submit" id="submitBtn" class="btn btn-primary">
                Save
              </button></div>
          </form>
        </div>
      </div>
      <!---->
      <div class="row" style="padding-top: 2rem;">
        <div class="col-8 mx-auto">
          <table aria-label="Recipes table" class="table table-striped table-hovering">
            <thead>
              <th scope="col"></th>
            </thead>
            <tbody></tbody>
          </table>
        </div>
      </div>
    </div>
  </div>
</div>
`;
